File: archive/pmoves_multi_agent_pro_pack/examples/python/basic/basic_client.py
#!/usr/bin/env python3
"""
Docling MCP Python Client - Basic Example

This example demonstrates basic usage of the Docling MCP service from Python.
It includes connection management, tool calling, and error handling.
"""

import asyncio
import json
import logging
import time
from typing import Dict, Any, Optional, List
from dataclasses import dataclass
from enum import Enum

import aiohttp
import sseclient


class OutputFormat(Enum):
    """Supported output formats."""
    MARKDOWN = "markdown"
    TEXT = "text"
    JSON = "json"


@dataclass
class MCPConfig:
    """Configuration for MCP client."""
    server_url: str = "http://localhost:3020/mcp"
    timeout: float = 30.0
    retry_attempts: int = 3
    retry_delay: float = 1.0
    max_connections: int = 10
    enable_logging: bool = True


class DoclingMCPClient:
    """Docling MCP Client with comprehensive features."""
    
    def __init__(self, config: MCPConfig = None):
        self.config = config or MCPConfig()
        self.session: Optional[aiohttp.ClientSession] = None
        self.is_connected = False
        self.request_id = 0
        self.pending_requests: Dict[str, asyncio.Future] = {}
        self.metrics = {
            'connections': 0,
            'requests': 0,
            'errors': 0,
            'response_times': []
        }
        
        # Setup logging
        if self.config.enable_logging:
            logging.basicConfig(
                level=logging.INFO,
                format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
            )
        self.logger = logging.getLogger(__name__)
    
    async def connect(self) -> bool:
        """Connect to the MCP server."""
        try:
            self.logger.info(f"Connecting to MCP server at {self.config.server_url}")
            
            # Create session with connection pooling
            connector = aiohttp.TCPConnector(
                limit=self.config.max_connections,
                limit_per_host=self.config.max_connections
            )
            
            timeout = aiohttp.ClientTimeout(total=self.config.timeout)
            self.session = aiohttp.ClientSession(
                connector=connector,
                timeout=timeout
            )
            
            # Test connection with health check
            health = await self.health_check()
            if health:
                self.is_connected = True
                self.metrics['connections'] += 1
                self.logger.info("Successfully connected to MCP server")
                return True
            else:
                self.logger.error("Health check failed")
                return False
                
        except Exception as e:
            self.logger.error(f"Connection failed: {e}")
            self.metrics['errors'] += 1
            return False
    
    async def disconnect(self):
        """Disconnect from the server."""
        if self.session:
            await self.session.close()
            self.session = None
            self.is_connected = False
            self.logger.info("Disconnected from MCP server")
    
    async def _make_request(self, method: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Make a request to the MCP server."""
        if not self.is_connected:
            raise ConnectionError("Not connected to server")
        
        self.request_id += 1
        request_id = str(self.request_id)
        
        # Create request
        request = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
            "params": params
        }
        
        # Create future for response
        future = asyncio.Future()
        self.pending_requests[request_id] = future
        
        try:
            start_time = time.time()
            self.metrics['requests'] += 1
            
            # Note: This is a simplified implementation
            # In a real implementation, you'd need to handle SSE properly
            # For now, we'll use HTTP POST to the MCP gateway
            
            if self.session:
                async with self.session.post(
                    self.config.server_url.replace('/mcp', '/tools/call'),
                    json=request
                ) as response:
                    if response.status == 200:
                        result = await response.json()
                        response_time = (time.time() - start_time) * 1000
                        self.metrics['response_times'].append(response_time)
                        return result
                    else:
                        raise Exception(f"HTTP {response.status}: {await response.text()}")
            else:
                raise ConnectionError("No active session")
                
        except Exception as e:
            self.metrics['errors'] += 1
            self.logger.error(f"Request failed: {e}")
            raise
        finally:
            self.pending_requests.pop(request_id, None)
    
    async def list_tools(self) -> List[Dict[str, Any]]:
        """List available tools."""
        result = await self._make_request("tools/list", {})
        return result.get('tools', [])
    
    async def call_tool(self, tool_name: str, arguments: Dict[str, Any] = None) -> Dict[str, Any]:
        """Call a specific tool."""
        params = {
            "name": tool_name,
            "arguments": arguments or {}
        }
        return await self._make_request("tools/call", params)
    
    async def health_check(self) -> Dict[str, Any]:
        """Perform health check."""
        try:
            return await self.call_tool("health_check")
        except Exception as e:
            self.logger.error(f"Health check failed: {e}")
            return {"status": "unhealthy", "error": str(e)}
    
    async def get_config(self) -> Dict[str, Any]:
        """Get server configuration."""
        return await self.call_tool("get_config")
    
    async def convert_document(self, file_path: str, output_format: OutputFormat = OutputFormat.MARKDOWN) -> str:
        """Convert a document to the specified format."""
        result = await self.call_tool("convert_document", {
            "file_path": file_path,
            "output_format": output_format.value
        })
        
        # Extract content from result
        if isinstance(result, dict) and 'content' in result:
            content = result['content']
            if isinstance(content, list) and len(content) > 0:
                return content[0].get('text', '')
            elif isinstance(content, str):
                return content
        
        return str(result)
    
    async def process_batch(self, file_paths: List[str], output_format: OutputFormat = OutputFormat.MARKDOWN) -> str:
        """Process multiple documents in batch."""
        result = await self.call_tool("process_documents_batch", {
            "file_paths": file_paths,
            "output_format": output_format.value
        })
        
        # Extract content from result
        if isinstance(result, dict) and 'content' in result:
            content = result['content']
            if isinstance(content, list) and len(content) > 0:
                return content[0].get('text', '')
            elif isinstance(content, str):
                return content
        
        return str(result)
    
    def get_metrics(self) -> Dict[str, Any]:
        """Get client metrics."""
        avg_response_time = (
            sum(self.metrics['response_times']) / len(self.metrics['response_times'])
            if self.metrics['response_times'] else 0
        )
        
        return {
            'connections': self.metrics['connections'],
            'requests': self.metrics['requests'],
            'errors': self.metrics['errors'],
            'avg_response_time_ms': round(avg_response_time, 2),
            'response_time_count': len(self.metrics['response_times'])
        }
    
    def reset_metrics(self):
        """Reset metrics."""
        self.metrics = {
            'connections': 0,
            'requests': 0,
            'errors': 0,
            'response_times': []
        }


async def main():
    """Main example function."""
    # Configure client
    config = MCPConfig(
        server_url="http://localhost:3020/mcp",
        timeout=30.0,
        retry_attempts=3,
        retry_delay=1.0
    )
    
    client = DoclingMCPClient(config)
    
    try:
        # Connect to server
        print("Connecting to Docling MCP server...")
        connected = await client.connect()
        
        if not connected:
            print("Failed to connect to server")
            return
        
        # Health check
        print("\nPerforming health check...")
        health = await client.health_check()
        print(f"Health status: {json.dumps(health, indent=2)}")
        
        # List available tools
        print("\nListing available tools...")
        tools = await client.list_tools()
        print(f"Available tools: {json.dumps(tools, indent=2)}")
        
        # Get server configuration
        print("\nGetting server configuration...")
        config = await client.get_config()
        print(f"Server config: {json.dumps(config, indent=2)}")
        
        # Example: Convert a document (you'll need actual files)
        # print("\nConverting document...")
        # result = await client.convert_document('/path/to/document.pdf', OutputFormat.MARKDOWN)
        # print(f"Conversion result: {result[:200]}...")
        
        # Get metrics
        print("\nClient metrics:")
        metrics = client.get_metrics()
        print(json.dumps(metrics, indent=2))
        
    except Exception as e:
        print(f"Error: {e}")
    
    finally:
        await client.disconnect()


if __name__ == "__main__":
    # Run the example
    asyncio.run(main())

File: archive/pmoves_multi_agent_pro_pack/examples/javascript/nodejs/basic-client.js
#!/usr/bin/env node
/**
 * Docling MCP Node.js Client - Basic Example
 * 
 * This example demonstrates basic usage of the Docling MCP service from Node.js.
 * It includes connection management, tool calling, and error handling.
 */

const EventSource = require('eventsource');
const fs = require('fs').promises;
const path = require('path');

class DoclingMCPClient {
    constructor(serverUrl, options = {}) {
        this.serverUrl = serverUrl;
        this.options = {
            timeout: options.timeout || 30000,
            retryAttempts: options.retryAttempts || 3,
            retryDelay: options.retryDelay || 1000,
            ...options
        };
        
        this.eventSource = null;
        this.isConnected = false;
        this.messageHandlers = new Map();
        this.requestId = 0;
        this.pendingRequests = new Map();
        this.metrics = {
            connections: 0,
            requests: 0,
            errors: 0,
            responseTimes: []
        };
        
        this.logger = options.logger || console;
    }

    /**
     * Connect to the MCP server
     */
    async connect() {
        return new Promise((resolve, reject) => {
            let attempts = 0;
            
            const tryConnect = () => {
                attempts++;
                this.logger.log(`Attempting to connect to ${this.serverUrl} (attempt ${attempts}/${this.options.retryAttempts})`);
                
                try {
                    this.eventSource = new EventSource(this.serverUrl);
                    
                    this.eventSource.onopen = () => {
                        this.isConnected = true;
                        this.metrics.connections++;
                        this.logger.log('Successfully connected to Docling MCP server');
                        resolve();
                    };

                    this.eventSource.onmessage = (event) => {
                        this.handleMessage(event.data);
                    };

                    this.eventSource.onerror = (error) => {
                        this.isConnected = false;
                        this.metrics.errors++;
                        this.logger.error('Connection error:', error);
                        
                        if (attempts < this.options.retryAttempts) {
                            this.logger.log(`Retrying connection in ${this.options.retryDelay}ms...`);
                            setTimeout(tryConnect, this.options.retryDelay);
                        } else {
                            reject(new Error(`Failed to connect after ${attempts} attempts`));
                        }
                    };

                } catch (error) {
                    this.logger.error('Connection failed:', error.message);
                    
                    if (attempts < this.options.retryAttempts) {
                        this.logger.log(`Retrying connection in ${this.options.retryDelay}ms...`);
                        setTimeout(tryConnect, this.options.retryDelay);
                    } else {
                        reject(error);
                    }
                }
            };
            
            tryConnect();
        });
    }

    /**
     * Disconnect from the server
     */
    disconnect() {
        if (this.eventSource) {
            this.eventSource.close();
            this.isConnected = false;
            this.logger.log('Disconnected from server');
        }
    }

    /**
     * Handle incoming messages
     */
    handleMessage(data) {
        try {
            const message = JSON.parse(data);
            
            if (message.id && this.pendingRequests.has(message.id)) {
                const { resolve, reject, startTime } = this.pendingRequests.get(message.id);
                this.pendingRequests.delete(message.id);
                
                const responseTime = Date.now() - startTime;
                this.metrics.responseTimes.push(responseTime);
                
                if (message.error) {
                    reject(new Error(message.error.message || 'Unknown error'));
                } else {
                    resolve(message.result);
                }
            }
        } catch (error) {
            this.logger.error('Error handling message:', error);
        }
    }

    /**
     * Call a tool on the MCP server
     */
    async callTool(toolName, toolArguments = {}) {
        if (!this.isConnected) {
            throw new Error('Not connected to server');
        }

        const requestId = ++this.requestId;
        const request = {
            jsonrpc: '2.0',
            id: requestId,
            method: 'tools/call',
            params: {
                name: toolName,
                arguments: toolArguments
            }
        };

        return new Promise((resolve, reject) => {
            const startTime = Date.now();
            this.metrics.requests++;
            
            this.pendingRequests.set(requestId, { resolve, reject, startTime });
            
            // Send request (Note: EventSource is read-only, this is a limitation)
            // In a real implementation, you'd need to use a different approach
            this.logger.warn('EventSource is read-only. In a real implementation, use WebSocket or HTTP POST');
            
            // Timeout handling
            setTimeout(() => {
                if (this.pendingRequests.has(requestId)) {
                    this.pendingRequests.delete(requestId);
                    reject(new Error('Request timeout'));
                }
            }, this.options.timeout);
        });
    }

    /**
     * List available tools
     */
    async listTools() {
        return this.callTool('list_tools');
    }

    /**
     * Health check
     */
    async healthCheck() {
        return this.callTool('health_check');
    }

    /**
     * Get server configuration
     */
    async getConfig() {
        return this.callTool('get_config');
    }

    /**
     * Convert a document
     */
    async convertDocument(filePath, outputFormat = 'markdown') {
        return this.callTool('convert_document', {
            file_path: filePath,
            output_format: outputFormat
        });
    }

    /**
     * Process documents in batch
     */
    async processBatch(filePaths, outputFormat = 'markdown') {
        return this.callTool('process_documents_batch', {
            file_paths: filePaths,
            output_format: outputFormat
        });
    }

    /**
     * Get client metrics
     */
    getMetrics() {
        const avgResponseTime = this.metrics.responseTimes.length > 0 
            ? this.metrics.responseTimes.reduce((a, b) => a + b, 0) / this.metrics.responseTimes.length 
            : 0;
        
        return {
            connections: this.metrics.connections,
            requests: this.metrics.requests,
            errors: this.metrics.errors,
            avgResponseTime: Math.round(avgResponseTime),
            responseTimes: this.metrics.responseTimes
        };
    }

    /**
     * Reset metrics
     */
    resetMetrics() {
        this.metrics = {
            connections: 0,
            requests: 0,
            errors: 0,
            responseTimes: []
        };
    }
}

/**
 * Example usage
 */
async function main() {
    const client = new DoclingMCPClient('http://localhost:3020/mcp', {
        timeout: 30000,
        retryAttempts: 3,
        retryDelay: 1000
    });

    try {
        // Connect to the server
        await client.connect();
        
        // Health check
        console.log('Performing health check...');
        const health = await client.healthCheck();
        console.log('Health check result:', health);
        
        // List available tools
        console.log('\nListing available tools...');
        const tools = await client.listTools();
        console.log('Available tools:', JSON.stringify(tools, null, 2));
        
        // Get server configuration
        console.log('\nGetting server configuration...');
        const config = await client.getConfig();
        console.log('Server configuration:', JSON.stringify(config, null, 2));
        
        // Example: Convert a document (you'll need to provide actual file paths)
        // console.log('\nConverting document...');
        // const result = await client.convertDocument('/path/to/document.pdf', 'markdown');
        // console.log('Conversion result:', result);
        
        // Get metrics
        console.log('\nClient metrics:');
        console.log(JSON.stringify(client.getMetrics(), null, 2));
        
    } catch (error) {
        console.error('Error:', error.message);
    } finally {
        client.disconnect();
    }
}

// Run the example if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = DoclingMCPClient;